### PR TITLE
cephfs-top: fix exceptions on small/large sized windows

### PR DESCRIPTION
**Pull request by joscollin**
* Fixes `exception: curses function returned NULL` when the window width is larger than expected.
Reproduced this issue on a 32 inch horizontal monitor. Tested the fix on 32 inch horizontal + 24 inch vertical monitor(s) (dual).
* Fixes `exception: addwstr() returned ERR`  when the window width is smaller than expected.

Fixes: https://tracker.ceph.com/issues/67859

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
jenkins test make check arm64

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
jenkins test api

**Review by dparmar18 — APPROVED** [review]

File: src/tools/cephfs/top/cephfs-top [hunk 1]
```diff
@@ -1163,7 +1172,10 @@ class FSTop(FSTopBase):
             elif cmd == curses.KEY_END:
                 hscrollOffset = self.PAD_WIDTH - self.viewportWidth - 1
             elif cmd == curses.KEY_RESIZE:
-                # terminal resize event. Update the viewport dimensions
+                # terminal resize event.
+                # Update the pad dimensions
+                self.PAD_WIDTH = DEFAULT_PAD_WIDTH
```
Comment (dparmar18) [inline src/tools/cephfs/top/cephfs-top:1177]: just a quick question, so when the terminal window is resized, the default pad width is 300 so does it also work with quite small screens?

**Comment by neesingh-rh** [comment]
@joscollin I looked into the second issue earlier through the BZ https://bugzilla.redhat.com/show_bug.cgi?id=2272580, and had this fix. Anyways, thanks for adding this fix into this PR. I have linked the BZ with the tracker for this PR.
If there's no issue, should I assign the respective BZ to you, as you have the PR for the fix?

**Review by neesingh-rh — APPROVED** [review]

**Comment by joscollin** [comment]
> @joscollin I looked into the second issue earlier through the BZ https://bugzilla.redhat.com/show_bug.cgi?id=2272580, and had this fix. Anyways, thanks for adding this fix into this PR. I have linked the BZ with the tracker for this PR. If there's no issue, should I assign the respective BZ to you, as you have the PR for the fix?

Please assign it to me, I wasn't aware that the BZ exists. Thanks @neesingh-rh

File: src/tools/cephfs/top/cephfs-top [hunk 2]
```diff
@@ -1138,7 +1154,13 @@ class FSTop(FSTopBase):
...
                         num_client = len(client_metadata)
                     vscrollEnd += num_client
-                    if self.create_header(stats_json, help, screen_title, 2):
+                    try:
+                        header_create = self.create_header(stats_json, help, screen_title, 2)
+                    except curses.error:
+                        curses.endwin()
+                        sys.stderr.write('Error creating header\n')
+                        exit()
+                    if header_create:
```
Comment (dparmar18) [inline src/tools/cephfs/top/cephfs-top:1163]: How about creating a helper func for this? `handle_header()`?

Comment (joscollin) [reply src/tools/cephfs/top/cephfs-top]: fixed, thanks

**Comment by joscollin** [comment]
Fixed the tox failures.

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
jenkins test make check

**Comment by joscollin** [comment]
@vshankar Work is done with this. Please do a review and QA.

**Comment by joscollin** [comment]
jenkins test make check arm64

**Review by vshankar — APPROVED** [review]
LGTM. I assume this has been manually verified since there is no way to test this in teuthology, yes? @joscollin

**Comment by joscollin** [comment]
Yes, its manually verified, as mentioned in the PR description. No issues
found.

**Comment by joscollin** [comment]
> LGTM. I assume this has been manually verified since there is no way to test this in teuthology, yes? @joscollin

@vshankar Can you merge this then?

**Comment by joscollin** [comment]
@vshankar ping

**Comment by vshankar** [comment]
> > LGTM. I assume this has been manually verified since there is no way to test this in teuthology, yes? @joscollin
> 
> @vshankar Can you merge this then?

yes. For some reason your comment via email did't show up as a notification.

**Pull request merged**